discovery/openstack/openstack.go

```diff
@@ -0,0 +1,202 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openstack
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/util/strutil"
+)
+
+const (
+	openstackLabel               = model.MetaLabelPrefix + "openstack_"
+	openstackLabelInstanceID     = openstackLabel + "instance_id"
+	openstackLabelInstanceName   = openstackLabel + "instance_name"
+	openstackLabelInstanceState  = openstackLabel + "instance_state"
+	openstackLabelInstanceFlavor = openstackLabel + "instance_flavor"
+	openstackLabelTag            = openstackLabel + "tag_"
+)
+
+var (
+	openstackSDRefreshFailuresCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "prometheus_sd_openstack_refresh_failures_total",
+			Help: "The number of OpenStack-SD scrape failures.",
+		})
+	openstackSDRefreshDuration = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Name: "prometheus_sd_openstack_refresh_duration_seconds",
+			Help: "The duration of a OpenStack-SD refresh in seconds.",
+		})
+)
+
+func init() {
+	prometheus.MustRegister(openstackSDRefreshFailuresCount)
+	prometheus.MustRegister(openstackSDRefreshDuration)
+}
+
+// Discovery periodically performs OpenStack-SD requests. It implements
+// the TargetProvider interface.
+type Discovery struct {
+	os       *gophercloud.AuthOptions
+	region   string
+	interval time.Duration
+	port     int
+}
+
+// NewDiscovery returns a new OpenStackDiscovery which periodically refreshes its targets.
+func NewDiscovery(conf *config.OpenstackSDConfig) (*Discovery, error) {
+	opts := gophercloud.AuthOptions{
+		IdentityEndpoint: conf.IdentityEndpoint,
+		Username:         conf.Username,
+		UserID:           conf.UserID,
+		Password:         conf.Password,
+		TenantName:       conf.TenantName,
+		TenantID:         conf.TenantID,
+		DomainName:       conf.DomainName,
+		DomainID:         conf.DomainID,
+	}
+
+	return &Discovery{
+		os:       &opts,
+		region:   conf.Region,
+		interval: time.Duration(conf.RefreshInterval),
+		port:     conf.Port,
+	}, nil
+}
+
+// Run implements the TargetProvider interface.
+func (d *Discovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	ticker := time.NewTicker(d.interval)
+	defer ticker.Stop()
```

Comment:
I'd move these two lines down to where the ticker is first used, like in `gce.go`. It also means that you only start the ticker after the first fetch completes, which is probably what you want.